### ISSUE TITLE
FOQUS Cloud SDOE Investigation

### ISSUE THREAD

**Issue by boverhof** [CCSI-Toolset/FOQUS#991]
# FOQUS Cloud SDOE 
- Do we have a story where SDOE usage in FOQUS flowsheet and interact with other components in a workflow?
- Do we have a story where SDOE runs existing FOQUS components?
- Do we have a story where SDOE can be parallelized and significant runtime improvements and/or up scaling?
- Can we save the candidate models on AWS S3 and make them available via cloud?

## SDOE
- Use cases:   Load files in pandas data frame, python functions from USF, etc.  Now using numpy, 
- Idea is run SDOE python code in AWS Lambda via AWS API Gateway
- Parallelize at the SDOE Random Sample Candidates level
- Hopefully start with runtimes of SDOE examples on order of hours.  Maybe NUSF.  Pedro looking into.

**Comment by sotorrio1** [comment]
I ran a few tests trying to get some numbers regarding possible scenarios where we can leverage running on AWS. I tried to get the worst case scenario for the candidate sets I used.

- **USF**: 
_candidate file to use_: examples/tutorial_files/SDOE/candidate.csv
_min design size_: 2
_max design size_: 10
_number of random starts_: 10^7 or 10^6
_estimated time_: 23 h 54 min 41 s or 2 h 23 min 28 s

In both cases we can parallelize the runs by design size and reduce runtime by 9x

- **NUSF**: 
_candidate file to use_: examples/tutorial_files/SDOE/SDoE_CCSI_example.csv
_design size_: 10
_scaling_: Direct
_MWR_: 5, 10, 20, 40, 60
_number of random starts_: 1,000
_estimated time_: ~5 h

We can parallelize by MWR value and reduce runtime by 5x

- **IRSF**: 
_candidate file to use_: examples/tutorial_files/SDOE/irsf-example1-candset.csv
_design size_: 10
_number of random starts_: 500
_estimated time_: ~5 h

We can parallelize by weight value and reduce runtime by 5x

**Comment by sotorrio1** [comment]
@boverhof what level of detail do you need? 
- Name of functions to run?
- name of UI elements for the function arguments? 
- input files?

**Comment by boverhof** [comment]
@sotorrio1 I'd like an example that I can run and prove out the value proposition by increasing design size and random starts.  I could try running the examples above on the cloud or on a local deployment, but I'm not sure I quite have enough information here to proceed.  Maybe a couple FOQUS screen shots of how to set this up?

Questions:
- What is the benefit of increasing random starts?  In layman's terms what does it mean?
- Explain the meaning of Design Size.  Can it just be bumped up to create additional generated designs?

**Comment by sotorrio1** [comment]
@boverhof Got it. I'll put together an example for one of the cases (perhaps Uniform Space-Filling - USF) with some screen shots. 

To answer your questions, I'll try my best but I'm also mentioning @mirandamartin so she can assist from the design of experiments perspective. 

1. The number of random starts determines the amount of times you run your optimization search of design points. The larger the number of random starts, the better your created design will be. 
2. Design size means the number of points you pick from the entire candidate set to create your final design. In the case of USF, since we specify a minimum design size and a maximum design size, we will be creating each one of those design sizes. In the NUSF or IRSF cases, we would need to call the function several times with different design sizes to create different designs.
3. MWR means maximum weight ratio. This one I'm going to leave @mirandamartin to explain it :)

**Comment by sotorrio1** [comment]
@boverhof please find below some instructions to run a USF design:

1. In the **SDoE module** make sure **Space-filling DoE (SDoE)** radio button is selected. 
2. Load the candidate set _examples/tutorial_files/SDOE/candidate.csv_ using the **Load Existing Set** button. 
3. Click **Continue**.
4. In the **Design Construction** panel click on the **Open SDoE Dialog** button. 

<img width="1438" alt="Screen Shot 2022-01-20 at 12 56 07 PM" src="https://user-images.githubusercontent.com/33531919/150420867-cb24e055-d209-4119-ad08-ef9af422e92e.png">

5. Increase **Max Design Size** to 10 or whatever number you want. SDoE will generate all the designs between min and max design sizes. 
6. Click **Estimate Runtime** to get a time estimate of the run time.
7. Increase the **Number of Random Starts** to 10<sup>6</sup> or 10<sup>7</sup>. You will see how the estimated runtime will increase. Note that I'm using 10<sup>3</sup> for this example to run it faster.
8. Click **Run SDoE**
9. The created designs will start populating the table on the left.

<img width="1440" alt="Screen Shot 2022-01-20 at 1 01 50 PM" src="https://user-images.githubusercontent.com/33531919/150421640-7454a58f-5092-4151-bcc6-54e157f63c6e.png">

Let me know if you need anything else or if you prefer an example for the other two cases (NUSF and/or IRSF).

**Comment by mirandamartin** [comment]
@boverhof Pedro did a good job of answering your questions I think, but I'll add in a few points. 

1. The basic idea of the number of random starts is indeed the higher the random starts, the better a design you'll get generally speaking. Let's say we choose 100 random starts. Each random start is a new separate run of the optimization algorithm trying to find the best design it can. So we'll get 100 separately determined "best" designs. Then FOQUS compares each of these 100 designs, chooses the 1 best design of these, and that 1 overall best design is returned. So, more random starts means more separately determined "best" designs to choose from at the end, and thus a better chance of finding a really good design. 

2. Just to reiterate what Pedro said, design size doesn't mean the number of designs FOQUS returns. It's the number of points (or rows in an excel sheet) that are included in the 1 design FOQUS returns. In USF, users can choose to find the 1 best design for each of the design sizes selected (say, 5, 6, and 7). So FOQUS would separately determine the 1 best design of size 5, the 1 best design of size 6, and the 1 best design of size 7, and return those 3 designs. Does that make sense? 

3. Maximum Weight Ratio is just a scaling factor for one of the columns in the candidate set, so it shouldn't  impact computation time. But to answer your question on what is MWR, it's a user-specified value that can change how the design points are dispersed through the region of interest (Do we want equal spacing? ... Do we want all the design points huddled in one corner of the input space?). It only applies when creating NUSF designs. 
![image](https://user-images.githubusercontent.com/55799605/150426804-c1717c7d-1798-40e9-ae6d-e79659e3d14b.png)

Did this help to answer your questions?
